### PR TITLE
NearestNeighborInterpolator

### PR DESCRIPTION
**Pull request by reedessick**
this patch implements the `NearestNeighborInterpolator` based on the sparsity-inducing assumptions from

> Abhirup Datta, Sudipto Banerjee, Andrew O. Finley & Alan E. Gelfand (2016) Hierarchical Nearest-Neighbor Gaussian Process Models for Large Geostatistical Datasets, Journal of the American Statistical Association, 111:514, 800-812, DOI: 10.1080/01621459.2015.1044091

This patch also refactors some of the inner workings of the `Interpolator` object in order to simplify inheritance.

**To Do**:

  * [ ] add support for parallelization (through mulitprocessing.pool?) for `_2diag`
    - should be embarrassingly parallelizable
  * [ ] implement `condition`
    - update logic to group target_x inot sets with the same neighbors
    - then invert cov_src_src one for each set of neighbors
    - make sure this plays nicely within `_2diag` and and parallelization thereof (each separate job handles one neighbor set and all it's associated target_x)
  * [ ] implement `compress`
    - need to compute the "full precision matrix" efficiently
  * [ ] implement `predict`
    - need to compute the off-diagonal elements of the "full cov matrix"
  * [ ] figure out a way to make the implementation jax-compatible
    - all I really need is to make `_2diag` jax compatible?

### CODE REVIEW

**Comment by reedessick** [comment]
To Do:

  * [ ] add support for parallelization (through mulitprocessing.pool?) for `_2diag`
    - should be embarrassingly parallelizable
  * [ ] implement `condition`
  * [ ] implement `compress` and `predict`

**Comment by reedessick** [comment]
using the uiuc table as a test, I measured the average runtime for calls to `loglikelihood` for different interpolators

  * GP = `mgpi.Interpolator.loglikelihood(...)` : the standard GP algorithm
    - Expected to scale as O(N**3)
  * NNGP = `mgpi.NearestNeighborInterpolator.loglikelihood(...)` : the NNGP algorithm
    - likelihood evaluation is expected to scale as O(N), but this data includes the cost of sorting the data into neighbor sets, which ends up being O(N**2) with our current implementation
    - note that the cost of constructing neighbor sets is subdominat at low sample sizes but eventually dominates for the largest training sets
  * (presorted)NNGP = `mgpi.NearestNeighborInterolator.loglikelihood(..., neighbors=neighbors)` : the NNG algorithm
    - this data is based on only the likelihood evaluation; we pre-sort the training data into neighbor sets once and reuse those sets
    - this should be indicative of the actual runtime during calls to `optimize_kernel` and `sample_kernel`

Markers (sometimes with error bars) denot the measured runtimes on my laptop. The dashed lines correspond to the expected scalings (listed above)

uiuc-test-nngp-timing-plot

As we can see, my implementation of the NNGP algorithm suffers significant overhead compared to the standard algorithm for low sample sizes. This is likely because I've implemented the NNGP with python loops whereas the standard algorithm can take advantage of numpy's broadcasting through calls to `numpy.linalg`. 

Nonetheless, for very large training sets, which are likely to be of interest to the community, we observe signficant runtime reductions by using the NNGP implementation. In fact, we also save in memory. The memory requirements for the standard GP algorithm grow so quickly that I could not preform tests on my laptop for the largest sample sizes.
